### PR TITLE
fixed error messages and initialization of subjob_cluster in workflow…

### PR DESCRIPTION
**Pull request by jannitzbon**
I started working with the workflow manager and came across two issues in `src/esm_runscripts/workflow.py`

- First, some error messages about unknown clusters could not be displayed correctly (lines 135 and 159) because of a mistake in the string formatting. I fixed these lines by providing the (hopefully) correct reference.

- Second, after fixing the error messages, I found another issue related to why they were called in the first place. In `initial_total_workflow`, the subjobs `prepcompute`, `compute` and `tidy` are only initialized if the field `subjobs` is not included in the `workflow` config section of the runscript. However, these should be added by default (according to the documentation (https://esm-tools.readthedocs.io/en/latest/workflow.html#). I suggest to fix this by addings these three jobs with the default settings if they are not provided with different settings in the `subjob` field of the `workflow` section:

```
    if not "subjobs" in config["general"]["workflow"]:
        config["general"]["workflow"]["subjobs"] = prepcompute
        config["general"]["workflow"]["subjobs"].update(compute)
        config["general"]["workflow"]["subjobs"].update(tidy)
    else:
        if not "prepcompute" in config["general"]["workflow"]["subjobs"]:
            config["general"]["workflow"]["subjobs"].update(prepcompute)
        if not "compute" in config["general"]["workflow"]["subjobs"]:
            config["general"]["workflow"]["subjobs"].update(compute)
        if not "tidy" in config["general"]["workflow"]["subjobs"]:
            config["general"]["workflow"]["subjobs"].update(tidy) 
```
After this fix, I was able to initialize an iterative coupling run.

### CODE REVIEW

**Review by pgierz — COMMENTED** [review]
From me only a comment, Miguel is far deeper in the workflow manager code than I am, so please wait for his green light.

I see no problems with this fix, aside from we need testing for the whole feature. 

And, one giant and over the top high five for Jan! This is exactly what we need and want for the project: a user tried something, found an error, fixed it, and then told us. +60000 points.

File: src/esm_runscripts/workflow.py
Comment (mandresm) [inline src/esm_runscripts/workflow.py:339]: I'm a bit puzzled about this, because we have 2 other offline coupled setups and we never found the problem you are describing. The fix looks totally reasonable to me, but the fact that you need it and others didn't makes me wonder what has been done differently. Can you share your runscripts and branch of ESM-Tools that you were using?

**Review by mandresm — COMMENTED** [review]
Hi @jannitzbon, thanks so much for your merge request. This is really helpful and the way to a sustainable development of the tools :)

I only have one minor concern, but after my question is answered I'll be merging this.

File: src/esm_runscripts/workflow.py
Comment (jannitzbon) [reply src/esm_runscripts/workflow.py:339]: Hi @mandresm , thanks for your quick response on this. In fact, I was wondering the same, since I followed the documentation and the example for the [pism-vilma](https://github.com/esm-tools/esm_tools/tree/release/runscripts/vilma-pism) coupling. Maybe some who is running those 2 setups you mentioned could help here?

I am using the branch `feature/cryogrid` which I just updated.

The runscripts can be found in `/albedo/home/jnitzbon/scripts/runscripts/awiesm-cryogrid` . Below, I copied the sections related to the workflow manger:

"General" runscript:
```
general:
    iterative_coupling: True
    base_dir: /albedo/work/user/jnitzbon/experiments/awiesm-cryogrid-2.1v1.0.0/
    use_venv: False
    coupling_script_dir: /albedo/home/jnitzbon/git/esm_tools/cryogrid-1.0.0/coupling/

model1:
    setup_name: awiesm
    runscript: runscript_awiesm_coupled_albedo.yaml
    chunk_size: 1
    chunk_unit: "years"

model2:
    setup_name: cryogrid
    runscript: runscript_cryogrid_coupled_albedo.yaml
    chunk_size: 1
    chunk_unit: "years"
```
AWI-ESM runscript:
```
[...]
awiesm:
    [...]
    workflow:
        next_run_triggered_by: couple_out
        subjobs: 
            couple_in:
                nproc: 1
                batch_or_shell: shell
                order_in_cluster: sequential
                run_before: prepcompute
                script_dir: ${general.coupling_script_dir} #${cryogrid.model_dir}/coupling/
                script: "scope_receive_echam.sh ${scope.config_file}"
                run_only: first_run_in_chunk
                skip_chunk_number: 1
            couple_out:
                nproc: 1
                batch_or_shell: shell
                order_in_cluster: sequential
                run_after: tidy
                script_dir: ${general.coupling_script_dir} #${cryogrid.model_dir}/coupling/
                script: "scope_send_echam.sh ${scope.config_file}"
                run_only: last_run_in_chunk
[...]
   ```
CryoGrid runscript:
```
[...]
cryogrid:
    [...]
    workflow:
        next_run_triggered_by: couple_out
        subjob_clusters:
            couple_in: couple_in
            couple_out: couple_out
        subjobs: 
            couple_in:
                nproc: 1
                batch_or_shell: shell
                order_in_cluster: sequential
                run_before: prepcompute
                script_dir: ${general.coupling_script_dir} #${cryogrid.model_dir}/coupling/
                script: "scope_receive_cryogrid.sh ${scope.config_file}"
                run_only: first_run_in_chunk
            couple_out:
                nproc: 1
                batch_or_shell: shell
                order_in_cluster: sequential
                run_after: tidy
                script_dir: ${general.coupling_script_dir} #${cryogrid.model_dir}/coupling/
                script: "scope_send_cryogrid.sh ${scope.config_file}"
                run_only: last_run_in_chunk
[...]
```

Comment (jannitzbon) [reply src/esm_runscripts/workflow.py:339]: One difference between my runscripts and the ones for the [awicm-pism](https://github.com/esm-tools/esm_tools/tree/release/runscripts/awicm-pism) coupling I noticed is that I added the workflow section within the awiesm section (which is already an online coupled setup) while it is put in the echam and fesom sections (which are standalone models) in the awicm-pism setup.

Comment (mandresm) [reply src/esm_runscripts/workflow.py:339]: That might be it, because coupled setups sections are merged into `general` during the parsing. I'll have a deeper look later on the day.

Comment (mandresm) [reply src/esm_runscripts/workflow.py:339]: Hi @jannitzbon, I finally got some time to look into this. I don't have the right permissions for `/albedo/home/jnitzbon/scripts/runscripts/awiesm-cryogrid`. Can you grant them or copy the scripts somewhere where I can access them?

Comment (jannitzbon) [reply src/esm_runscripts/workflow.py:339]: Hi @mandresm , which permissions do you need? I added write access for the runscripts, please check again. 

I also moved the runscripts here: `/albedo/work/user/jnitzbon/temp/`

Comment (mandresm) [reply src/esm_runscripts/workflow.py:339]: I only needed read permissions. However, I'm still not able to access the runscripts:
```shell
$ ls /albedo/work/user/jnitzbon/temp/
ls: cannot access '/albedo/work/user/jnitzbon/temp/': Permission denied
```

Maybe it's because I don't have reading access to the parent directories?

If you prefer it you can copy the runscripts here `/albedo/work/user/mandresm/shared_with_others`

Comment (jannitzbon) [reply src/esm_runscripts/workflow.py:339]: Okay. I could not ls or cp to your work directory either, so I put the scripts here:

`/albedo/work/projects/paleo_work/jnitzbon/awiesm-cryogrid`

**Comment by mandresm** [comment]
Thanks @jannitzbon, using your runscript, the commit previous to this fix, and moving the `workflow` into `awiesm` I manged to reproduce the problem and gain a better understanding of what was going on. Now I'm confident that this can be merged without generating problems to other configurations.

#approve-changes

**Review by mandresm — APPROVED** [review]

**Comment by mandresm** [comment]
#bump

**Pull request merged**
